### PR TITLE
IGNITE-23650 SQL Calcite: Refactor IndexScan/TableScan classes hierarchy

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        release();
+    }
+
+    /** */
+    private synchronized void reserve() {
+        if (reserved != null)
+            return;
+
+        GridDhtPartitionTopology top = cctx.topology();
+        top.readLock();
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java:87]: minor: let's use a line break

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        if (!done || !(topFut.topologyVersion().compareTo(topVer) >= 0
+            && cctx.shared().exchange().lastAffinityChangedTopologyVersion(topFut.initialVersion()).compareTo(topVer) <= 0)) {
+            top.readUnlock();
+
+            throw new ClusterTopologyException("Topology was changed. Please retry on stable topology.");
+        }
+
+        List<GridDhtLocalPartition> toReserve;
+
+        if (cctx.isReplicated()) {
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java:102]: Let's use line breaks below

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/TreeIndexIterable.java [hunk 3]
```diff
@@ -154,9 +114,7 @@ private void advance() throws IgniteCheckedException {
                 return;
 
             while (next == null && cursor.next()) {
-                IdxRow idxRow = cursor.get();
-
-                next = indexRow2Row(idxRow);
+                next = cursor.get();
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/TreeIndexIterable.java:117]: The brackets aren't required any more

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import org.apache.ignite.internal.processors.affinity.AffinityTopologyVersion;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.cache.distributed.dht.GridDhtTopologyFuture;
+import org.apache.ignite.internal.processors.cache.distributed.dht.topology.GridDhtLocalPartition;
+import org.apache.ignite.internal.processors.cache.distributed.dht.topology.GridDhtPartitionState;
+import org.apache.ignite.internal.processors.cache.distributed.dht.topology.GridDhtPartitionTopology;
+import org.apache.ignite.internal.util.typedef.F;
+
+/** */
+public abstract class AbstractCacheScan<Row> implements Iterable<Row>, AutoCloseable {
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java:34]: Do we need this class ? Used only `AbstractCacheColumnsScan<Row> extends AbstractCacheScan<Row`.

Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java:34]: It's preparation for https://issues.apache.org/jira/browse/IGNITE-23562, to minimize changes by this ticket.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java:88]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/AbstractCacheScan.java:103]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/TreeIndexIterable.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/TreeIndexIterable.java:117]: Fixed

**Review by Vladsz83 — APPROVED** [review]

**Pull request closed without merging**